scripts/changelog-check

```diff
@@ -2,25 +2,50 @@
 require 'open3'
 require 'optparse'
 
-CHANGELOG_REGEX = %r{^\+- (?<category>[\w -]{2,}): (?<change>.+) \((?<pr>(?:#\d+, ?)*(?:#\d+))\)$}
-# A valid diff line is in the form of:
-# +- CATEGORY: CHANGE_DESCRIPTION (#PULL_REQUEST_NUMBER)
-def is_valid_changelog_diff_addition?(line)
-  matches = CHANGELOG_REGEX.match(line)
-  !matches.nil?
+CHANGELOG_REGEX =
+  %r{^changelog: (?<category>[\w -/]{2,}), (?<subcategory>[A-Z][\w -]{2,}), (?<change>.+)$}
+CATEGORIES = [
+  'Improvements', 'Accessibility', 'Bug Fixes', 'Behind the Scenes Changes'
+]
+SKIP_CHANGELOG_MESSAGE = '[skip changelog]'
+
+SquashedCommit = Struct.new(:title, :commit_messages, keyword_init: true)
+ChangelogEntry = Struct.new(:category, :subcategory, :change, :pr_number, keyword_init: true)
+
+# A valid entry has a line in a commit message in the form of:
+# changelog: CATEGORY, SUBCATEGORY, CHANGE_DESCRIPTION
+def build_changelog(line)
+  CHANGELOG_REGEX.match(line)
 end
 
-def get_changelog_diff_lines(base_branch, source_branch)
-  diff, status = Open3.capture2(
-    'git', 'diff', '--merge-base', base_branch, source_branch,
-    'CHANGELOG.md'
+# Transforms a formatted git log into structured objects.
+# The git format ends up printing a single commit as:
+#
+# title: Remove unused IdV controller view (#5922)
+# body:**Why**: Because it's unused.
+# * Add changelog, change constant name
+# DELIMITER
+# The string is first split on DELIMITER, and then the body is split into
+# individual lines.
+def get_git_log(base_branch, source_branch)
+  format = '--pretty=title: %s%nbody:%b%nDELIMITER'
+  log, status = Open3.capture2(
+    'git', 'log', format, "#{base_branch}...#{source_branch}"
   )
-  raise 'git diff failed' unless status.success?
 
-  diff_lines = diff.chomp.split("\n")
-  index = diff_lines.find_index { |line| line.start_with?('@@') }
-  return [] if index.nil?
-  diff_lines[(index+1)..].filter { |line| !line.start_with?('@@') && line != ' ' }
+  raise 'git log failed' unless status.success?
+
+  log.strip.split('DELIMITER').map { |commit|
+    commit.split("\nbody:").map do |commit_message_lines|
+      commit_message_lines.split(%r{[\r\n]}).filter { |line| line != '' }
```

Comment:
This is the least comprehensible and most brittle part, I don't like it very much, so I am extremely interested in suggestions or improvements here.

Reply:
Just this line? Or the whole thing?

Reply:
Whole thing feels very terse to me. Maybe it's less important as long as the input and output are clear-ish

Reply:
can always add unit tests to make yourself feel more confident!